mock-tikv/src/server.rs

```diff
@@ -0,0 +1,494 @@
+        _req: kvproto::kvrpcpb::DeleteRangeRequest,
+        _sink: grpcio::UnarySink<kvproto::kvrpcpb::DeleteRangeResponse>,
+    ) {
+        todo!()
+    }
+
+    fn raw_get(
+        &mut self,
+        ctx: grpcio::RpcContext,
+        req: kvproto::kvrpcpb::RawGetRequest,
+        sink: grpcio::UnarySink<kvproto::kvrpcpb::RawGetResponse>,
+    ) {
+        let mut resp = RawGetResponse::default();
+        if let Some(v) = self.inner.raw_get(req.get_key()) {
+            resp.set_value(v);
+        } else {
+            resp.set_not_found(true);
+        }
+        spawn_unary_success!(ctx, req, resp, sink);
+    }
+
+    fn raw_batch_get(
+        &mut self,
+        ctx: grpcio::RpcContext,
+        req: kvproto::kvrpcpb::RawBatchGetRequest,
+        sink: grpcio::UnarySink<kvproto::kvrpcpb::RawBatchGetResponse>,
+    ) {
+        let mut resp = kvproto::kvrpcpb::RawBatchGetResponse::default();
+        resp.set_pairs(self.inner.raw_batch_get(req.get_keys()));
+        spawn_unary_success!(ctx, req, resp, sink);
+    }
+
+    fn raw_put(
+        &mut self,
+        ctx: grpcio::RpcContext,
+        req: kvproto::kvrpcpb::RawPutRequest,
+        sink: grpcio::UnarySink<kvproto::kvrpcpb::RawPutResponse>,
+    ) {
+        self.inner
+            .raw_put(req.get_key().to_vec(), req.get_value().to_vec());
+        let resp = RawPutResponse::default();
+        spawn_unary_success!(ctx, req, resp, sink);
+    }
+
+    fn raw_batch_put(
+        &mut self,
+        ctx: grpcio::RpcContext,
+        req: kvproto::kvrpcpb::RawBatchPutRequest,
+        sink: grpcio::UnarySink<kvproto::kvrpcpb::RawBatchPutResponse>,
+    ) {
+        let pairs = req.get_pairs();
+        self.inner.raw_batch_put(pairs);
+        let resp = RawBatchPutResponse::default();
+        spawn_unary_success!(ctx, req, resp, sink);
+    }
+
+    fn raw_delete(
+        &mut self,
+        ctx: grpcio::RpcContext,
+        req: kvproto::kvrpcpb::RawDeleteRequest,
+        sink: grpcio::UnarySink<kvproto::kvrpcpb::RawDeleteResponse>,
+    ) {
+        let key = req.get_key();
+        let res = self.inner.raw_delete(key);
+        let mut resp = RawDeleteResponse::default();
+        if res.is_none() {
+            resp.set_error("Key not exist".to_owned());
+        }
+        spawn_unary_success!(ctx, req, resp, sink);
+    }
+
+    fn raw_batch_delete(
+        &mut self,
+        ctx: grpcio::RpcContext,
+        req: kvproto::kvrpcpb::RawBatchDeleteRequest,
+        sink: grpcio::UnarySink<kvproto::kvrpcpb::RawBatchDeleteResponse>,
+    ) {
+        let keys: &[Vec<u8>] = req.get_keys();
+        let res = self.inner.raw_batch_delete(keys);
+        let mut resp = RawBatchDeleteResponse::default();
+        if res.is_err() {
+            resp.set_error(format!(
+                "Non-existent keys:[{}]",
+                res.err()
+                    .unwrap()
+                    .iter()
+                    .map(|k| std::str::from_utf8(k).unwrap())
```

Comment:
You can't convert the bytes into  string since the bytes might not be valid unicode. You could use the `debug` method or write a custom function to get a string representation.

Did you check TiKV's behaviour when trying batch delete non-existent keys? Does it give an error?

Reply:
I asked Yilin. Deleting non-existent keys should not return error.